### PR TITLE
Queues - Anna Barklund -    ride-share-two

### PR DESCRIPTION
**Pull request by amb54**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |  There were three classes to create that needed to communicate with each other. The csv file for the trip class showed information of id related to all three classes. My decision was to have the Trip class as a connector for the three classes.      |
| Describe a concept that you gained more clarity on as you worked on this assignment. | The idea  how the classes are communicating with each other.   |
| Describe a _nominal test_ that you wrote for this assignment. |  For the trip class method self.all I tested that all 600 lines in the csv file created instances.  |
| Describe an _edge case test_ that you wrote for this assignment. |  Testing of the accepted phone numbers in the Rider class initialize method. They must be in the format  NPA-NXX-xxxx.  N cannot be 0 or 1.  XX cannot be 11  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  I like it. It makes it more clear. I get a simple overview of the project, everything is not spinning around in my head. I get more focused. I really like the tests : ) |

### CODE REVIEW

File: specs/driver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      driver1.average_rating.must_be_kind_of Float
+    end
+
+    it "Test of calculation of the average of the rating with number of rides > 0  " do
+      sum_of_ratings = driver1.list_of_trips.map { |trip| trip.rating}.sum.to_f
+      driver1.average_rating.must_equal sum_of_ratings/9
+    end
+
+    it "Test of calculation of the average of the rating with number of rides == 0  " do
+      driver100 = RideSharing::Driver.find(100)
```
Comment (droberts-sea) [inline specs/driver_spec.rb:159]: Because you defined `driver100` in the let block above, you don't need to re-find it here.

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        all_drivers << self.new(driver_hash)
+      end
+      return all_drivers
+    end
+
+    def self.find(driver_id)
+      found_driver = self.all.select { |driver| driver.id == driver_id}
+      begin
+        raise ArgumentError.new("Id number #{driver_id} does not exist") if found_driver == []
+      rescue ArgumentError => exception
```
Comment (droberts-sea) [inline lib/driver.rb:51]: Typically you don't `rescue` an exception that you yourself `raise`. Instead, `raise`-ing an exception is used to communicate to whoever called this method that there was a problem, and `rescue` is used if some method `raise`s an exception you know how to handle.

Comment (droberts-sea) [reply lib/driver.rb:51]: In fact, you don't really even need to `puts` anything here - returning `nil` is sufficient to let the caller know that there was no `Driver` with that ID.

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    def list_of_trips
+      RideSharing::Trip.find_all_trips_for_driver(@id)
+    end
+
+    def average_rating
+      if list_of_trips != []
+        return list_of_trips.map { |trip| trip.rating}.sum.to_f/ list_of_trips.length
+      else
+        puts "#{@name} with id##{@id} has not yet taken any trips.\nHence the average rating cannot be calcualted"
+        return Float::NAN
```
Comment (droberts-sea) [inline lib/driver.rb:66]: As above, you don't need to `puts` here - returning `NAN` is sufficient.

File: lib/rider.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      end
+      return found_rider.first
+    end
+
+    def list_of_trips
+      RideSharing::Trip.find_all_trips_for_rider(@id)
+    end
+
+    def previous_drivers
+      drivers = list_of_trips.map { |trip| trip.find_driver}.delete_if {|driver| driver == nil}.uniq { |driver| driver.id}
```
Comment (droberts-sea) [inline lib/rider.rb:62]: Good use of enumerables here!

All of these are clear in what they do, but it still works out to a pretty long line. Something like

```ruby
drivers = list_of_trips.map { |trip| trip.find_driver}
drivers.delete_if! {|driver| driver == nil}
drivers.uniq! { |driver| driver.id}
```

might be a little easier to read.

File: lib/rider.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      return drivers
+    end
+
+    private
```
Comment (droberts-sea) [inline lib/rider.rb:66]: Nice use of `private` to hide methods that aren't part of your API.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   yes - good work!	|
|   Answer comprehension questions	|   yes	|
|  **Driver** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   yes |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   yes |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips method to retrieve the associated drivers | yes |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |   yes |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   yes |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   yes |
|   Created a method to retrieve all trips by driver id |    yes |
|   Created a method to retrieve all trips by rider id |     yes |
|   **Overall**	|   	|

Great work overall! This code is clean and readable, your tests cover everything I'm looking for and your git habits appear strong. Keep up the good work.
